src/sparse_merkle_tree/smt.rs

```diff
@@ -1349,6 +1350,108 @@ impl SparseMerkleTree {
+            next_query.hash = [query.hash.as_slice(), sibling_hash]
+                .concat()
+                .hash_with_kind(HashKind::Branch);
+        } else {
+            next_query.hash = [sibling_hash, query.hash.as_slice()]
+                .concat()
+                .hash_with_kind(HashKind::Branch);
+        }
+        next_query.slice_bitmap();
+
+        next_query
+    }
+
+    fn is_bitmap_valid(
+        sibling: &QueryProofWithProof,
+        query: &QueryProofWithProof,
+    ) -> Result<bool, SMTError> {
+        let is_sibling_empty = utils::is_empty_hash(&sibling.hash);
+        if (is_sibling_empty && query.binary_bitmap[0])
+            || (!is_sibling_empty && !query.binary_bitmap[0])
+        {
+            return Err(SMTError::InvalidInput(String::from(
+                "bitmap is not consistent with the nodes type",
+            )));
+        }
+        let is_query_empty = utils::is_empty_hash(&query.hash);
+        if (is_query_empty && sibling.binary_bitmap[0])
+            || (!is_query_empty && !sibling.binary_bitmap[0])
+        {
+            return Err(SMTError::InvalidInput(String::from(
+                "bitmap is not consistent with the nodes type",
+            )));
+        }
+        if !utils::array_equal_bool(&query.binary_bitmap[1..], &sibling.binary_bitmap[1..]) {
+            return Err(SMTError::InvalidInput(String::from(
+                "nodes do not share common path",
+            )));
+        }
+
+        Ok(true)
+    }
+
+    fn prepare_result_proof(
+        proof: &Proof,
+        adding_sibling_hashes: &[(usize, Vec<u8>)],
+        removing_sibling_hashes: &NestedVec,
+        removing_keys: &[&[u8]],
+        next_sibling_hash_index: usize,
+    ) -> Result<Proof, SMTError> {
+        let mut new_sibling_hashes = proof.sibling_hashes.clone();
+        if next_sibling_hash_index != proof.sibling_hashes.len() {
+            return Err(SMTError::InvalidInput(String::from(
+                "Not all sibling hashes were used",
+            )));
+        }
+        for hash in adding_sibling_hashes {
+            if new_sibling_hashes.contains(&hash.1) {
+                return Err(SMTError::InvalidInput(String::from(
+                    "Duplicate sibling hashes",
+                )));
+            }
+            if hash.0 > new_sibling_hashes.len() {
+                new_sibling_hashes.push(hash.1.clone());
+            } else {
+                new_sibling_hashes.insert(hash.0, hash.1.clone());
+            }
+        }
+
+        let mut temp_sibling_hashes: NestedVec = vec![];
+        for hash in new_sibling_hashes.iter() {
+            if !removing_sibling_hashes.contains(hash) {
+                temp_sibling_hashes.push(hash.clone());
+            }
+        }
+        if new_sibling_hashes.len() - temp_sibling_hashes.len() != removing_sibling_hashes.len() {
+            return Err(SMTError::InvalidInput(String::from(
+                "All removed sibling are not in sibling hashes",
```

Comment:
```suggestion
                "All removed siblings are not in sibling hashes",
```

Reply:
Should we rephrase this sentence? It's quite hard to understand what it actually means.